turbopack/crates/turbo-tasks-backend/src/utils/dash_map_multi.rs

```diff
@@ -39,94 +36,123 @@ impl<K: Eq + Hash, V, S: BuildHasher> RefMut<'_, K, V, S> {
+
+    // Use `determine_shard` instead of `determine_map` to avoid extra rehashing.
+    // This u64 -> usize conversion also happens internally within DashMap using `as usize`.
+    // See: `DashMap::hash_usize`
+    let s1 = map.determine_shard(h1 as usize);
+    let s2 = map.determine_shard(h2 as usize);
+
+    let eq1 = |other: &(K, _)| key1.eq(&other.0);
+    let eq2 = |other: &(K, _)| key2.eq(&other.0);
+
     let shards = map.shards();
     if s1 == s2 {
         let mut guard = shards[s1].write();
-        let e1 = guard
-            .raw_entry_mut()
-            .from_key(&key1)
-            .or_insert_with(|| (key1.clone(), SharedValue::new(insert_with())));
-        let mut key1_ptr = e1.0 as *const K;
-        let mut value1_ptr = e1.1.get_mut() as *mut V;
-        let key2_ptr;
-        let value2_ptr;
-        match guard.raw_entry_mut().from_key(&key2) {
-            hash_map::RawEntryMut::Occupied(e) => {
-                let e2 = e.into_key_value();
-                key2_ptr = e2.0 as *const K;
-                value2_ptr = e2.1.get_mut() as *mut V;
-            }
-            hash_map::RawEntryMut::Vacant(e) => {
-                let e2 = e.insert(key2.clone(), SharedValue::new(insert_with()));
-                key2_ptr = e2.0 as *const K;
-                value2_ptr = e2.1.get_mut() as *mut V;
-                // inserting a new entry might invalidate the pointers of the first entry
-                let e1 = guard.get_key_value_mut(&key1).unwrap();
-                key1_ptr = e1.0 as *const K;
-                value1_ptr = e1.1.get_mut() as *mut V;
-            }
-        }
+
+        // we need to call `find_or_find_insert_slot` to avoid overwriting existing entries, but we
+        // can't use the returned bucket until after we get `bucket2` (below)
+        let _ = guard
+            .find_or_find_insert_slot(h1, eq1, hash_entry)
+            .unwrap_or_else(|slot| unsafe {
+                // SAFETY: This slot was previously returned by `find_or_find_insert_slot`, and no
+                // mutation of the table has occured since that call.
+                guard.insert_in_slot(h1, slot, (key1.clone(), SharedValue::new(insert_with())))
+            });
+
+        let bucket2 = guard
+            .find_or_find_insert_slot(h2, eq2, hash_entry)
+            .unwrap_or_else(|slot| unsafe {
+                // SAFETY: See previous call above
+                guard.insert_in_slot(h2, slot, (key2.clone(), SharedValue::new(insert_with())))
+            });
+
+        // Getting `bucket2` might invalidate the bucket pointer of the first entry, *even if no
+        // insert happens* as `RawTable::find_or_find_insert_slot` will *sometimes* resize the
+        // table, as it unconditionally reserves space for a potential insertion.
+        let bucket1 = guard.find(h1, eq1).expect(
+            "failed to find bucket of previously inserted item, is the hash or eq implementation \
+             incorrect?",
+        );
```

Comment:
I could do an ugly hack where I call `.reserve(2)` at the top of this conditional block before `find_or_find_insert_slot` to avoid this second lookup, but two elements being in the same shard is an unlikely branch to begin with (in reality we use many shards), so I think it's better to do this the safer way.